loguru/_locks_machinery.py

```diff
@@ -2,13 +2,45 @@
 import threading
 import weakref
 
+
+class HandlerLockNotReentrant(RuntimeError):
+    pass
+
+
+class _HandlerLockWrapper:
+    """Lock wrapper for the handler.
+
+    This avoids thread-local deadlock by propagating an exception rather than waiting forever.
+    See: https://github.com/Delgan/loguru/issues/712
+    """
+
+    def __init__(self, lock):
+        self._lock = lock
+        self._local = threading.local()
+        self._local.acquired = False
+
+    def __enter__(self):
+        if getattr(self._local, "acquired", False):
+            raise HandlerLockNotReentrant(
+                "Tried to emit a log in the process of handling a log. Log handler is not re-entrant."
+            )
+        self._local.acquired = True
+        self._lock.acquire()
+        return
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._lock.release()
+        self._local.acquired = False
+        return
+
+
 if not hasattr(os, "register_at_fork"):
 
     def create_logger_lock():
         return threading.Lock()
 
     def create_handler_lock():
-        return threading.Lock()
+        return _HandlerLockWrapper(threading.Lock())
```

Comment:
The approach here changes the behaviour for the handler lock in general. So this applies not only to `Handler.emit`, but also to `Handler.stop`, and `Handler.complete_async`.

I'm not sure if that is correct? If it is, do we need additional handling in those methods?

If this behaviour should only apply for `emit`, I think the solution shouldn't touch this file at all, and I will move it to `_handler.py`.

Reply:
Indeed, we should also protect the other methods using the `lock`, this is a very good idea!

However, I'm thinking it might be better to explicitly wrap the `lock` on use instead of on creation. I prefer to differentiate the context manager of the lock itself because in some cases it is not required. In particular, I'm thinking to [the `lock` inside the `_queued_writer` thread](https://github.com/Delgan/loguru/blob/08c9a871c706463a17142aaae90be0fd71e5a51f/loguru/_handler.py#L268) which does not need to be protected.

I'm thinking of replacing `with self._lock` with `with self._protected_lock():` for example. There is certainly a risk if we forget to use it later, but I'll keep an eye on it and I think it should be fine.

Reply:
👍